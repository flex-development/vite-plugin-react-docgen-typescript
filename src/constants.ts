/**
 * @file RDT Plugin - Constants
 * @module vite-plugin-react-docgen-typescript/constants
 */

/** @const {'vite:react-docgen-typescript'} PLUGIN_NAME - Plugin name */
export const PLUGIN_NAME = 'vite:react-docgen-typescript' as const
