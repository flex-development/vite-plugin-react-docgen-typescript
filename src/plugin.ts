/**
 * @file RDT Plugin - Implementation
 * @module vite-plugin-react-docgen-typescript/plugin
 */

import { createFilter } from '@rollup/pluginutils'
import MagicString from 'magic-string'
import path from 'node:path'
import * as rdt from 'react-docgen-typescript'
import type { TransformResult } from 'rollup'
import * as telejson from 'telejson'
import { PLUGIN_NAME } from './constants'
import type Options from './options'
import type PluginReactDocgenTypeScript from './plugin-type'

/**
 * Creates a `react-docgen-typescript` plugin.
 *
 * @see https://github.com/styleguidist/react-docgen-typescript
 * @see https://vitejs.dev/guide/api-plugin.html
 *
 * @param {Options} [options] - Plugin options
 * @return {PluginReactDocgenTypeScript} Vite `react-docgen-typescript` plugin
 */
const plugin = ({
  apply,
  componentNameResolver,
  customComponentTypes = [],
  enforce = 'pre',
  exclude = ['**/**.stories.tsx'],
  handler,
  include = ['**/**.tsx'],
  propFilter = prop => !prop.parent?.fileName.includes('node_modules'),
  savePropValueAsString = false,
  shouldExtractLiteralValuesFromEnum = false,
  shouldExtractValuesFromUnion = false,
  shouldIncludeExpression = false,
  shouldIncludePropTagMap = true,
  shouldRemoveUndefinedFromOptional = true,
  skipChildrenPropWithoutDoc = true,
  sourcemap = true,
  tsconfig = path.resolve('tsconfig.json')
}: Options = {}): PluginReactDocgenTypeScript => {
  /**
   * Filter function that determines which modules should be operated on.
   *
   * @see https://github.com/rollup/plugins/tree/master/packages/pluginutils
   *
   * @const {(id: unknown) => boolean} filter
   */
  const filter: (id: unknown) => boolean = createFilter(include, exclude)

  /**
   * Component docgen info parser.
   *
   * @see https://github.com/styleguidist/react-docgen-typescript#usage
   *
   * @const {rdt.FileParser} parser
   */
  const parser: rdt.FileParser = rdt.withCustomConfig(tsconfig, {
    componentNameResolver,
    customComponentTypes,
    propFilter,
    savePropValueAsString,
    shouldExtractLiteralValuesFromEnum,
    shouldExtractValuesFromUnion,
    shouldIncludeExpression,
    shouldIncludePropTagMap,
    shouldRemoveUndefinedFromOptional,
    skipChildrenPropWithoutDoc
  })

  return {
    apply,
    enforce,
    name: PLUGIN_NAME,
    transform: async (code: string, id: string): Promise<TransformResult> => {
      // do nothing if module should not be operated on
      if (!filter(id)) return

      /**
       * Component docgen info parsed from {@link id}.
       *
       * @see {@link rdt.ComponentDoc}
       *
       * @const {rdt.ComponentDoc[]} docs
       */
      const docs: rdt.ComponentDoc[] = parser.parse(id)

      // bail if missing component docgen info
      if (docs.length === 0) return null

      /**
       * {@link code} as `MagicString`.
       *
       * @see https://github.com/Rich-Harris/magic-string
       *
       * @const {MagicString} src
       */
      const src: MagicString = new MagicString(code)

      for (let doc of docs) {
        // apply additional processing to component docgen info
        if (handler) doc = await handler(doc, code, id)

        /**
         * {@link doc} stringified.
         *
         * @see https://github.com/storybookjs/telejson
         *
         * @const {string} __docgenInfo
         */
        const __docgenInfo: string = telejson.stringify(doc)

        // add __docgenInfo property
        src.append(`\n;${doc.displayName}.__docgenInfo=${__docgenInfo}`)
      }

      if (sourcemap === false) return src.toString()

      return {
        code: src.toString(),
        map: src.generateMap({
          ...(sourcemap === true ? { hires: true } : sourcemap),
          source: id
        })
      }
    }
  }
}

export default plugin
