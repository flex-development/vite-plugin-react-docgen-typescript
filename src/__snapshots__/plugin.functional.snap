// Vitest Snapshot v1

exports[`functional:plugin > should not transform module if docgen info is missing 1`] = `null`;

exports[`functional:plugin > should not transform module if id does not match include pattern 1`] = `undefined`;

exports[`functional:plugin > should not transform module if id matches exclude pattern 1`] = `undefined`;

exports[`functional:plugin > should transform module if id matches include pattern 1`] = `
"/**
 * @file Fixtures - Button
 * @module fixtures/Button
 */

import {
  forwardRef,
  useState,
  type ButtonHTMLAttributes,
  type ForwardedRef
} from 'react'
import s, { css } from 'styled-components'

interface ButtonProps extends ButtonHTMLAttributes<HTMLButtonElement> {
  /**
   * Button background color.
   *
   * @default 'black'
   */
  $bg?: ButtonHTMLAttributes<HTMLButtonElement>['color']

  /**
   * Button text color.
   *
   * @default 'white'
   */
  $color?: ButtonHTMLAttributes<HTMLButtonElement>['color']

  /**
   * Button font size.
   *
   * @default '1.25rem'
   */
  $fontsize?: number | string

  /**
   * Mark button as disabled.
   */
  disabled?: boolean
}

interface CounterButtonProps extends ButtonProps {
  /**
   * Initial click count.
   *
   * @default 0
   */
  count?: number
}

/**
 * Renders an HTML \`<button>\` element.
 *
 * @see https://developer.mozilla.org/docs/Web/HTML/Element/button
 * @see https://developer.mozilla.org/docs/Web/API/HTMLButtonElement
 *
 * @visibleName Button
 */
const Button = s.button.attrs<ButtonProps>(p => ({
  'aria-disabled': p.disabled ?? undefined,
  'aria-hidden': p.hidden ?? undefined
}))<ButtonProps>(p => {
  return css\`
    background-color: \${p.$bg};
    color: \${p.$color};
  \`
})

Button.defaultProps = {
  $bg: 'black',
  $color: 'white',
  $fontsize: '1.25rem',
  type: 'button'
}

/**
 * Renders a button that displays how many times it has been clicked.
 *
 * @param {CounterButtonProps} [props={}] - Component props
 * @param {ForwardedRef<HTMLButtonElement>} [ref] - \`<button>\` reference
 * @return {JSX.Element} Counter button markup
 */
const CounterButton = forwardRef(function (
  props: CounterButtonProps = {},
  ref?: ForwardedRef<HTMLButtonElement>
): JSX.Element {
  const { count, ...rest } = props

  const [clicks, setclicks] = useState<number>(count!)

  /**
   * Increases {@link clicks}.
   *
   * @return {void} Nothing when complete
   */
  const increment = (): void => setclicks(prev => prev + 1)

  return (
    <Button {...rest} onClick={increment} ref={ref}>
      {clicks}
    </Button>
  )
})

CounterButton.displayName = 'CounterButton'
CounterButton.defaultProps = {}

export {
  Button as default,
  CounterButton,
  type ButtonProps,
  type CounterButtonProps
}

;Button.__docgenInfo={\\"tags\\":{\\"see\\":\\"https://developer.mozilla.org/docs/Web/HTML/Element/button\\\\nhttps://developer.mozilla.org/docs/Web/API/HTMLButtonElement\\",\\"visibleName\\":\\"Button\\"},\\"filePath\\":\\"\\",\\"description\\":\\"Renders an HTML \`<button>\` element.\\",\\"displayName\\":\\"Button\\",\\"methods\\":[],\\"props\\":{\\"ref\\":{\\"defaultValue\\":null,\\"description\\":\\"\\",\\"name\\":\\"ref\\",\\"parent\\":\\"_undefined_\\",\\"declarations\\":[],\\"required\\":false,\\"type\\":{\\"name\\":\\"((instance: HTMLButtonElement | null) => void) | RefObject<HTMLButtonElement> | null\\"},\\"tags\\":{}},\\"$bg\\":{\\"defaultValue\\":{\\"value\\":\\"black\\"},\\"description\\":\\"Button background color.\\",\\"name\\":\\"$bg\\",\\"parent\\":{\\"fileName\\":\\"\\",\\"name\\":\\"ButtonProps\\"},\\"declarations\\":[],\\"required\\":false,\\"type\\":{\\"name\\":\\"string\\"},\\"tags\\":{\\"default\\":\\"'black'\\"}},\\"$color\\":{\\"defaultValue\\":{\\"value\\":\\"white\\"},\\"description\\":\\"Button text color.\\",\\"name\\":\\"$color\\",\\"parent\\":{\\"fileName\\":\\"\\",\\"name\\":\\"ButtonProps\\"},\\"declarations\\":[],\\"required\\":false,\\"type\\":{\\"name\\":\\"string\\"},\\"tags\\":{\\"default\\":\\"'white'\\"}},\\"$fontsize\\":{\\"defaultValue\\":{\\"value\\":\\"1.25rem\\"},\\"description\\":\\"Button font size.\\",\\"name\\":\\"$fontsize\\",\\"parent\\":{\\"fileName\\":\\"\\",\\"name\\":\\"ButtonProps\\"},\\"declarations\\":[],\\"required\\":false,\\"type\\":{\\"name\\":\\"string | number\\"},\\"tags\\":{\\"default\\":\\"'1.25rem'\\"}},\\"theme\\":{\\"defaultValue\\":null,\\"description\\":\\"\\",\\"name\\":\\"theme\\",\\"parent\\":\\"_undefined_\\",\\"declarations\\":[],\\"required\\":false,\\"type\\":{\\"name\\":\\"any\\"},\\"tags\\":\\"_duplicate_[\\\\\\"props\\\\\\",\\\\\\"ref\\\\\\",\\\\\\"tags\\\\\\"]\\"},\\"as\\":{\\"defaultValue\\":null,\\"description\\":\\"\\",\\"name\\":\\"as\\",\\"parent\\":\\"_undefined_\\",\\"declarations\\":[],\\"required\\":false,\\"type\\":{\\"name\\":\\"undefined\\"},\\"tags\\":\\"_duplicate_[\\\\\\"props\\\\\\",\\\\\\"ref\\\\\\",\\\\\\"tags\\\\\\"]\\"},\\"forwardedAs\\":{\\"defaultValue\\":null,\\"description\\":\\"\\",\\"name\\":\\"forwardedAs\\",\\"parent\\":\\"_undefined_\\",\\"declarations\\":[],\\"required\\":false,\\"type\\":{\\"name\\":\\"undefined\\"},\\"tags\\":\\"_duplicate_[\\\\\\"props\\\\\\",\\\\\\"ref\\\\\\",\\\\\\"tags\\\\\\"]\\"}}}
;CounterButton.__docgenInfo={\\"tags\\":{\\"param\\":\\"props - Component props\\\\nref - \`<button>\` reference\\",\\"return\\":\\"Counter button markup\\"},\\"filePath\\":\\"\\",\\"description\\":\\"Renders a button that displays how many times it has been clicked.\\",\\"displayName\\":\\"CounterButton\\",\\"methods\\":[],\\"props\\":{\\"count\\":{\\"defaultValue\\":{\\"value\\":\\"0\\"},\\"description\\":\\"Initial click count.\\",\\"name\\":\\"count\\",\\"parent\\":{\\"fileName\\":\\"\\",\\"name\\":\\"CounterButtonProps\\"},\\"declarations\\":[],\\"required\\":false,\\"type\\":{\\"name\\":\\"number\\"},\\"tags\\":{\\"default\\":\\"0\\"}},\\"$bg\\":{\\"defaultValue\\":{\\"value\\":\\"black\\"},\\"description\\":\\"Button background color.\\",\\"name\\":\\"$bg\\",\\"parent\\":{\\"fileName\\":\\"\\",\\"name\\":\\"ButtonProps\\"},\\"declarations\\":[],\\"required\\":false,\\"type\\":{\\"name\\":\\"string\\"},\\"tags\\":{\\"default\\":\\"'black'\\"}},\\"$color\\":{\\"defaultValue\\":{\\"value\\":\\"white\\"},\\"description\\":\\"Button text color.\\",\\"name\\":\\"$color\\",\\"parent\\":{\\"fileName\\":\\"\\",\\"name\\":\\"ButtonProps\\"},\\"declarations\\":[],\\"required\\":false,\\"type\\":{\\"name\\":\\"string\\"},\\"tags\\":{\\"default\\":\\"'white'\\"}},\\"$fontsize\\":{\\"defaultValue\\":{\\"value\\":\\"1.25rem\\"},\\"description\\":\\"Button font size.\\",\\"name\\":\\"$fontsize\\",\\"parent\\":{\\"fileName\\":\\"\\",\\"name\\":\\"ButtonProps\\"},\\"declarations\\":[],\\"required\\":false,\\"type\\":{\\"name\\":\\"string | number\\"},\\"tags\\":{\\"default\\":\\"'1.25rem'\\"}},\\"disabled\\":{\\"defaultValue\\":null,\\"description\\":\\"Mark button as disabled.\\",\\"name\\":\\"disabled\\",\\"parent\\":{\\"fileName\\":\\"\\",\\"name\\":\\"ButtonProps\\"},\\"declarations\\":[],\\"required\\":false,\\"type\\":{\\"name\\":\\"boolean\\"},\\"tags\\":{}}}}"
`;
